### ISSUE TITLE
DefendLane

### ISSUE THREAD

**Issue by Nostrademous** [Nostrademous/Dota2-FullOverwrite#192]
- [ ] Don't use TP if we can walk there (I saw Lina that was by safe-lane jungle shrine, TP to mid, to defend it. She could have walked there in the amount of time it took to TP, approx.

- [ ] Once we arrive, bots are too weary to engage against stronger enemy defense and just hang back and let the tower fall, need to fix this (either don't defend if we won't contest, or bring more firepower and contest). We should not use pure # of heroes as an estimate who to TP

- [ ] Once we decide to defend, we should probably push the wave out until 0.5 lane-amount is reached (to the river) before going back to do something else like farming/laning/pushing (unless it is very important, like we can get a kill or are needed in a team-fight)

**Comment by Keithenneu** [comment]
I agree to most of it, just some notes:
1. yeah. also, support should tp, not carries (unless necessary)
2. We should simply use the same determination we use for fighting. stalling the push is an option as well.
3. Ideally, that shold happen by itself, as they want to push, or farm the lane. We don't want to keep several core heroes busy just and only because a tower got pushed.

**Comment by Keithenneu** [comment]
BTW, I'll tinker with global map considerations for a few more days (I feel like we need a game plan, and stuff like that. Killing a tower because we happen to be there is nice. But I feel like we need some Blood-Lust in the game.)
If that goes horribly wrong I'll go back to bug fixing / improving stuff.

**Comment by Keithenneu** [comment]
Oh, and for 1. and 2.: The push detection is broken atm. number of heroes in 1200 range around the tower won't cut it. Improving that should stabilize the fight determination results.

**Comment by Keithenneu** [comment]
The bot's perception range is way to low in general. That's another thing I'm trying to take on in the map stuff.

**Comment by Nostrademous** [comment]
One other thing... we are still very bad about defense in our own base.

**Comment by Keithenneu** [comment]
Yeah, we just cant predict pushes at this time. And defending base with just the guys that happen to be around isn't going to work.

**Comment by Nostrademous** [comment]
Checked in a change where we actually "move towards" the tower we are to defend if we are too far away from it... we were just standing there at times in "defendlane" mode and far from tower.

**Comment by Riveon** [comment]
Another issue is that sometimes they TP to a tower when it isn't safe to do so. E.g. Viper and CM tp'd to the bottom tier 2 tower where there was an enemy Jugg bot (and another bot). Viper immediately retreated (meaning there was no point in Tping there instead of the tier 3 tower. (CM stayed in by herself and died.) 
This doesn't happen nearly as much as with the default bots though thankfully lol.
